src/controller/CHIPDeviceController.cpp

```diff
@@ -1025,6 +1028,114 @@ CHIP_ERROR DeviceCommissioner::UnpairDevice(NodeId remoteDeviceId)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR DeviceCommissioner::SendEchoRequest(NodeId remoteDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    Messaging::SendFlags sendFlags;
+    System::PacketBufferHandle payloadBuf;
+    char * requestData = nullptr;
+
+    uint32_t size = mPingParams.GetEchoReqSize();
+    VerifyOrExit(size <= kMaxPayloadSize, err = CHIP_ERROR_MESSAGE_TOO_LONG);
+
+    requestData = static_cast<char *>(chip::Platform::MemoryAlloc(size));
+    VerifyOrExit(requestData != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    snprintf(requestData, size, "Echo Message %" PRIu32 "\n", mPingParams.GetEchoReqCount() + 1);
+
+    payloadBuf = MessagePacketBuffer::NewWithData(requestData, size);
+    VerifyOrExit(!payloadBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    if (mPingParams.IsUsingMRP())
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNone);
+    }
+    else
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kNoAutoRequestAck);
+    }
+
+    ChipLogDetail(Controller, "Send echo request message with payload size: %d bytes to Node: %" PRIu64, size, remoteDeviceId);
+
+    err = mEchoClient.SendEchoRequest(std::move(payloadBuf), sendFlags);
+
+    if (err == CHIP_NO_ERROR)
+    {
+        mPingParams.SetWaitingForEchoResp(true);
+        mPingParams.IncrementEchoReqCount();
+    }
+
+exit:
+    if (requestData != nullptr)
+    {
+        chip::Platform::MemoryFree(requestData);
+    }
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "Send echo request failed, err: %s", ErrorStr(err));
+    }
+
+    return err;
+}
+
+CHIP_ERROR DeviceCommissioner::PingDevice(NodeId remoteDeviceId, uint32_t maxCount, uint32_t waitTimeMillis, uint32_t payloadSize,
+                                          bool usingMRP)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // TODO: temprary create a SecureSessionHandle from node id to unblock end-to-end test. Complete solution is tracked in
+    // issue:4451
+    err = mEchoClient.Init(mExchangeMgr, { remoteDeviceId, 0, 0 });
+    SuccessOrExit(err);
+
+    // Arrange to get a callback whenever an Echo Response is received.
+    mEchoClient.SetEchoResponseReceived(OnHandleEchoResponse);
+
+    mPingParams.Reset();
+    mPingParams.SetEchoInterval(waitTimeMillis);
+    mPingParams.SetEchoReqSize(payloadSize);
+    mPingParams.SetUsingMRP(usingMRP);
+
+    // Connection has been established. Now send the EchoRequests.
+    for (unsigned int i = 0; i < maxCount; i++)
+    {
+        err = SendEchoRequest(remoteDeviceId);
+
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Controller, "Send request failed: %s\n", ErrorStr(err));
+            break;
```

Comment:
you can goto-exit instead and place the echo client shutdown in the cleanup block.

Could we RAII the cleanup instead?

Reply:
If we go to exist, then we don't need to shutdown the mEchoClient. Yes, we can refactor the EchoClient with RAII, but I would like to leave this refactor in a separate PR.

Reply:
I would understand changing existing code to be out of scope for refactors - don't want to both add functionality and refactor in the same PR.

However this is new code, so I expect it to be withouth gotoexit to start with.

Reply:
Acked